server/permalinks.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v25/github"
+)
+
+const maxPermalinkReplacements = 10
+const permalinkReqTimeout = 2 * time.Second
+
+// replacement holds necessary info to replace github permalinks
+// in messages with a code preview block
+type replacement struct {
+	index      int               // index of the permalink in the string
+	word       string            // the permalink
+	captureMap map[string]string // named regex capture group of the link
+}
+
+// getReplacements returns the permalink replacements that needs to be performed
+// on a message.
+func (p *Plugin) getReplacements(msg string) []replacement {
+	// find the permalinks from the msg using a regex
+	matches := p.githubRegex.FindAllStringSubmatch(msg, -1)
+	indices := p.githubRegex.FindAllStringIndex(msg, -1)
+	replacements := make([]replacement, 0, maxPermalinkReplacements)
+	for i, m := range matches {
+		// have a limit on the no. of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// populate the captureMap with the extracted groups of the regex
+		for j, name := range p.githubRegex.SubexpNames() {
+			if j == 0 {
+				r.captureMap = make(map[string]string)
+				continue
+			}
```

Comment:
Could we move this initialization outside of the for loop?

Reply:
Sure, but we still need to `continue` if j == 0. But moving out the `captureMap` initialization is more readable. Done.